ocm-lib introduction (#367)
open-component-model/ocm: docs/lib/introduction.md

# Introduction to the **ocm-lib**

As the **ocm-lib** is a large and complex library, it can be rather frustrating to get started. This document is
supposed to support you in this process.

## Section 1 - Complexity

As mentioned in the [specification](https://github.com/open-component-model/ocm-spec#storage-technology), an
*OCM Repository* is an interpretation layer on top of existing storage technologies and is not itself a repository
technology.  
Therefore, the ocm-lib has to deal with multiple

> [storage technologies](https://github.com/open-component-model/ocm-spec#storage-technology) and their individuel

Reviewer: ```suggestion
[storage technologies](https://github.com/open-component-model/ocm-spec#storage-technology) and their individual
```

respective go types.

The [registry](introduction/prototype-based-typeregistry/registry/registry.go) has a function implementing the dynamic
unmarshaling logic. To understand this, it's best to check out the well-documented method within the example.

#### Section 2.1.2 - Factory-Based Type Registry

In _Factory-Based Type Registries_, the Registry is essentially a map that stores the model type name (key) and a
factory object (thus, an object capable of producing Message objects). Thereby, it is possible to overcome the
limitations of the Prototype-Based Type Registries (the necessity of a dedicated go type per model type and not being
able to edit the object after unmarshaling).

Based on the knowledge about Prototype-Based Type Registry, it should be possible to understand the well-documented
[Factory-Based Type Registry](introduction/factory-based-typeregistry) example.

#### Note

The ocm-lib uses has several kinds of type registries (thus, corresponding to our example, not only a Message Registry
but also for other types). There, it has a generic implementation of a registry which can be reused.  

> Furthermore, its implementation of the registry has to be thread safe and provides some convenience functionility.

Reviewer: ```suggestion
Furthermore, its implementation of the registry has to be thread safe and provides some convenience functionality.
```

Therefore, it may be slightly more difficult to understand than our examples.

> #### Section 2.1.3 - Purpose

Reviewer: I propose to move this "Purpose" Section up, maybe include this directly under "Section 2.1 - (Model) Type Registries". It is crucial to first of all understand (and therefore motivate the reader) _why_ these complex type registries are required in the OCM Lib, and only then talk about the _how_

[Factory-Based Type Registry](introduction/factory-based-typeregistry) example.

#### Note

The ocm-lib has several kinds of type registries (thus, corresponding to our example, not only a Message Registry
but also for other types). Therefore, it has a generic implementation of a registry which can be reused.  
Furthermore, its implementation of the registry has to be thread safe and provides some convenience functionality.
Therefore, it may be slightly more difficult to understand than our examples.

### Section 2.2 - Contexts

The ocm-lib defines multiple _Context_ types. The documentation of the library states that objects of these _Context_
types bundle all settings and extension point implementations for this area.

**Extension Point Implementations:** On a conceptual level, each package providing a type that implements an interface
that has a corresponding _Type Registry_ is considered an extension point implementation (in the previous
examples, the _simplemessage package_ and the _complexmessage package_ are extension points as they
implement the _Message_ interface and there is a _Message Type Registry_).  

> **Settings:** On a conceptual level, settings are values that may be set for specific objects or passed to specific

Reviewer: ```suggestion
**Settings:** Settings are values that may be set for specific objects or passed to specific
```

same
artifact (thus, it corresponds to a Namespace or rather OCI Repository in that regard), but implements the _OCI
Repository_ interface.

**OCM Repository Types:**
So, corresponding to OCI Repositories and OCI Repository Types, any _go type_ that implements the following
[interface](../../pkg/contexts/ocm/internal/repository.go) can be considered an _OCM Repository_:

```
type Repository interface {
   GetContext() Context

   GetSpecification() RepositorySpec
   ComponentLister() ComponentLister

   ExistsComponentVersion(name string, version string) (bool, error)
   LookupComponentVersion(name string, version string) (ComponentVersionAccess, error)
   LookupComponent(name string) (ComponentAccess, error)

   Close() error
}
```

Again, within the ocm-lib, there are multiple implementations of this interface. But most importantly, the package
[`genericocireg`](../../pkg/contexts/ocm/repositories/genericocireg) provides an implementation of the _OCM Repository
interface_ that is based on the _OCI Repository interface_. In other words, this package provides a mapping of the

> _OCM Repository_ functionility to the _OCI Repository_ functionality. The

Reviewer: ```suggestion
_OCM Repository_ functionality to the _OCI Repository_ functionality. The
```

[`ocireg`](../../pkg/contexts/ocm/repositories/ocireg) package and the [`ctf`](../../pkg/contexts/ocm/repositories/ctf)
merely provide the functionality to convert, or rather wrap, their _OCI Repository Spec_ to an _OCM Repository Spec_
(remember, spec types were pretty much serializable representations and factories for the actual type, as explained
[here](introduction/context/registry/registry.go)).  
For each _OCI Repository Type_, there also exists a corresponding _OCM Repository Type_.
Thus, as there is an **oci** _repository type_ `OCIRegistry`, there is also an **ocm** _repository type_ `OCIRegistry`,
as there is an **oci** _repository type_ `CommonTransportFormat`, there is also an **ocm** _repository type_
`CommonTransportFormat`.

> Within the ocm-lib this is implemented through an **addition to the factory-based type registry** concept. In case the

Reviewer: I guess a word is missing here after "In case the ... "

does not find an entry for a certain type (within the Decode method), this
[implementation (defaultScheme)](../../pkg/runtime/scheme.go) of a factory-based type registry has a _fallback factory_
(called `defaultdecoder`). For the OCM Repository Type Registry, this _fallback factory_ is usually
[initialized](../../pkg/contexts/ocm/builder.go) with a Type Registry itself, the
[`DefaultDelegationRegistry`](../../pkg/contexts/ocm/repositories/genericocireg/type.go) which is based on the

> [`delegationRegistry`](../../pkg/contexts/ocm/internal/delegation.go). This Type Registry's iterates through all its

Reviewer: ```suggestion
[`delegationRegistry`](../../pkg/contexts/ocm/internal/delegation.go). This Type Registry iterates through all its
```

> factories and checks whether they knew that type. Currently, the only factory known by this _fallback Type Registry_ is

Reviewer: ```suggestion
factories and checks whether on of them know that type. Currently, the only factory known by this _fallback Type Registry_ is
```

the one registered in the [`genericocireg`](../../pkg/contexts/ocm/repositories/genericocireg/type.go) package which
uses the **oci** _repository types registry_ to perform the unmarshaling (thus, if I registered the type in the
oci repository types registry, it is automatically available for the ocm repository registry).

**Other Repository Types:**
Besides _OCM_ and _OCI Repository Types_, there currently also exist a number of

> [_Credential Repository Types_](../../pkg/contexts/credentials/repositories) that have to implement the following

Reviewer: I don't quite get this: If these Credential Repo Types already exist, what is then meant with "that have to implement the following interface" ? If they do exist, wouldn't that mean they have already implemented the interface ?

[interface](../../pkg/contexts/credentials/internal/repository.go):

```
type Repository interface {
   ExistsCredentials(name string) (bool, error)
   LookupCredentials(name string) (Credentials, error)
   WriteCredentials(name string, creds Credentials) (Credentials, error)
}
```

### Section 3.2.2 - Access Type Scheme

An `AccessTypeScheme` is an object that maps a number of _access types (= model type)_ to corresponding
_factories_ providing an access spec object which is again a _factory_ for providing an object of a specific type or
with
a specific configuration for dealing with this _access type (= model type)_.

From the perspective of the ocm-lib, any _go type_ that implements the following

> [interface](../../pkg/contexts/ocm/internal/accesstypes.go) can be considered an _OCI Repository_:

Reviewer: Why can the existence of a go type implementing this AccessMethod interface be considered a OCI repository ? Below, an implementation for S3 or GitHub are mentioned as examples, which is not an OCI repositories, so maybe rather "OCM Repository" is meant here ?

Author: That's a copy-paste mistake. It should be ...can be considered an _Access Method_.

**CAREFUL - There are two conceptually relevant details here!**

1) There are **Global** and **Local** **Access Method types**. **Global types** (such as ociArtifact or github)
   implement the procedure to access the artifact blobs themselves. **Local types**, on the contrary, delegate the
   implementation of the procedure to the respective component repository (as each type of component repository may
   handle the storage of artifact blobs alongside the component descriptor differently).
2) **AccessMethods return the artifact blob as is**. Thus, if a helm chart is stored as an .tar, the AccessMethod will
   return the bytes of the .tar. If, on the other hand, the helm chart is stored as an tar.gz , the AccessMethod will
   return the bytes of the .tar.gz. Therefore, for further processing, the MimeType() can be read (inspecting the
   MimeType and conversion to a target format can be done with downloaders - a convenience feature implemented by the
   ocm library kind of as an add-on).

### Section 3.3 - Constraint Registries aka. Handler Registries (and Handler-Registration Handler Registries)

> **Constraint Registries aka. Handler Registries**  

Reviewer: Why are these called _CONSTRAINT_ Registries ? That term seems a bit unintuitive, no ?

specific configuration for dealing with this model type (e.g. a _repository_ or an _access method_).

From a technical perspective **Constraint Registries** work pretty much exactly the same. But their purpose is
not unmarshaling. While an _object_, or rather a _factory_, within a _Type Registry_ is selected within the decoding
logic based on the specified _type_ (a respective _Decode Spec_ method is typically exposed by the Type Registry
itself),
an _object_ within a _Constraint Registry_ is selected during some other event based on a number of constraints (a
respective _Lookup Handler_ method is typically exposed by the Constraint Registry itself). So instead of a _type_, e.g.
a _Blob Handler Registry_ (a _Blob Handler_ is a handler responsible for uploading blobs into a specific store) stores
a triple of _RepositoryType_ (e.g. `OCIRegistry`, `CommonTransportFormat`, ...), _ArtifactType_ (e.g. `OCIArtifact`,
`ociImage`, ...) and _MimeType_ (e.g. `text/plain`, `application/gzip`, ...). The upload logic passes the current
contextual information about the underlying _storage repository type_ and the _artifact_ and _mime type_ at hand to the

> _Lookup Handler_ method a _Constraint Registry_, which eventually returns a number of _Handlers_ fulfilling the

Reviewer: ```suggestion
_Lookup Handler_ method of a _Constraint Registry_, which eventually returns a number of _Handlers_ fulfilling the
```

[interface](../../pkg/registrations/registrations.go):

```
type HandlerRegistrationRegistry[T any, O any] interface {
   HandlerRegistrationHandler[T, O]
   RegisterRegistrationHandler(path string, handler HandlerRegistrationHandler[T, O])
   GetRegistrationHandlers(name string) []*RegistrationHandlerInfo[T, O]
}
```

Thus, a `HandlerRegistrationHandlerRegistry` also is itself also a  `HandlerRegistrationHandler`.

Without going into too much detail, the purpose of a _Handler-Registration Handler_ is to _register_ a _Handler_
identified by the `handler string` parameter at a _Constraint Registry_.  
Usually, directly within the library, such a functionality is not really necessary. If you want to add e.g. a new
_type_ of `BlobHandler`, you implement the `BlobHandler` interface and register that implementation at a corresponding
`BlobHandlerRegistry` by calling the BlobHandlerRegistry's _Register_ method (e.g. in an
[init](../../pkg/contexts/ocm/blobhandler/oci/ocirepo/blobhandler.go) function to register that implementation in a
DefaultRegistry). Thereby, you have to pass an object of that handler implementation to the _Register_ method.  

> But imagine you are working with the _ocm-cli_ and you want to register either a predefined `BlobHandler` (in other

Reviewer: ```suggestion
But imagine you are working with the _ocm-cli_ and you want to register a predefined `BlobHandler` (in other
```

UniformRepositorySpecs for [OCM Repositories](../../pkg/contexts/ocm/internal/uniform.go) and for
[OCI Repositories](../../pkg/contexts/oci/internal/uniform.go). Furthermore, _ocm_ and _oci references_, or rather
_reference strings_ (e.g. `eu.gcr.io/gardener-project/landscaper/examples/charts/hello-world:1.0.0`) are parsed into a
[representation that is based on a `UniformRepositorySpec`](../../pkg/contexts/oci/ref.go). Thus, to get access to the
respective Repository through the ocm-lib, this `UniformRepositorySpec` first has to be converted into a corresponding
`RepositorySpec`.  
In conclusion, a `RepositorySpecHandlers` _Constraint Registry_ is an object that maps a _repository type
(= constraint)_ to a number of corresponding `RepositorySpecHandler` objects that might be able to convert a
`UniformRepositorySpec` into the appropriate `RepositorySpec` for that _Repository Type_. Therefore, a
`RepositorySpecHandler` has to implement the following [interface](../../pkg/contexts/ocm/internal/uniform.go):

```
type RepositorySpecHandler interface {
   MapReference(ctx Context, u *UniformRepositorySpec) (RepositorySpec, error)
}
```

> A `RepositorySpecHandlers` Registry **is not** also a `HandlerRegistrationHandlerRegistry`.

Reviewer: Why is it necessary to state this so explicitly here ? And if it is necessary, some reasoning why this is not the case should be provided.

### Section 3.2.4 - Blob Handler Registry

The `BlobHandlerRegistry` can be considered a _Constraint Registry_. To avoid confusions, let's first clarify that the
name
_Blob Handler_ might be misleading as the only purpose of a _Blob Handler_ is the upload of blobs. This is also
immediately visible when looking at their [interface](../../pkg/contexts/ocm/internal/blobhandler.go):

```
type BlobHandler interface {
   StoreBlob(blob BlobAccess, artType, hint string, global AccessSpec, ctx StorageContext) (AccessSpec, error)
}
```

A much more suitable name would therefore be _Blob Uploader_.

When a _Component Version_, or rather, a _Component Descriptor_, is imported into an _OCM Repository_, a decision has
to be made about how to deal with the _artifacts_, or rather, the _artifact blobs_ (e.g. does it stay a local artifact
stored alongside the _Component Descriptor_ or is it uploaded into an _OCI Registry_ as an independent _OCI Artifact_?).
These decisions can be made within BlobHandlers. The `StoreBlob` method has to return an `AccessSpec` that can then be
inserted into the corresponding new _Component Descriptor_.

> A `BlobHanderRegistry` **is** also a `HandlerRegistrationHandlerRegistry`.

Reviewer: Same as above: If it is necessary to state this here so explicitly, then some reasoning would be nice.

digest may
change through unpacking and repacking processes (dates may be added or changed, the order of specific elements
may change). As a consequence, the same artifact could have different digests which would be a problem regarding
signing. Therefore, there exist _digesters_ for _specific mime types_ e.g. an OCI Artifact digester, that calculates the
digest over the byte sequence of the OCI artifact representation which always has to be the same per specification.
Generally, a `BlobDigester` is anything that allows to calculate the digest for blobs with specific mime types through
the following [interface](../../pkg/contexts/ocm/internal/digesthandler.go):

```
type BlobDigester interface {
   GetType() DigesterType
   DetermineDigest(resType string, meth AccessMethod, preferred signing.Hasher)   (*DigestDescriptor, error)
}
```

A BlobDigesterRegistry currently **is not** also a `HandlerRegistrationHandlerRegistry`.

## Section 3.4 - Contexts

Here, the different _Contexts_ shown in the overview are explained.

A context object is the _entry point_ for using dedicated functional areas. It bundles all _settings_ and

> _extensions point implementations_ for this area.

Reviewer: ```suggestion
_extension point implementations_ for this area.
```
